MSSQL: Handle updates that return multiple resultsets

jonstace-hsf opened the pull request:
This is very much a MSSQL-specific issue as it has to do with the interactions between the way MSSQL returns query results, ODBC and PyODBC.

Generally, when you run a query against MSSQL, using a tool such as SQL Server Management Studio, you get two output tabs, one is the 'results' and the other is 'messages'. A lot of the time, messages will show something like this:
```
(1 row affected)
```
However, if you're running an update against a table that has an update trigger on it, depending upon how that trigger is written, you'll get multiple rows like the above in the messages tab, one for the main update query and one for each update/insert/delete statement within that trigger. If you have PRINT statements in your triggers, or you're calling a stored procedure that has PRINT statements in it, then they will come back in the messages tab too.

The way that messages come back from the server is some deep, dark, low-level ODBC thing that I don't really want to have to try to understand but the important point is that PyODBC treats the data stream coming back from MSSQL as separate result sets, separated by the various messages. The outcome from all of this is that if you submit some SQL to PyODBC using Connection.execute() or Cursor.execute() then the query will stop running until Cursor.nextset() is called to get the next part of the results. 

In the context of Pynonymizer, I was running a before script to disable all my triggers as I didn't want them to run, but my before script had some print statements that listed the triggers being disabled. This resulted in not all triggers being disabled, because that before script just stopped on the first print statement. Then, all the subsequent updates from my config were running with the triggers enabled and only anonymizing some of the data.

This could be updated through documentation (e.g. to not use print statements in before and after scripts), or we could add 'SET NOCOUNT ON;' to the start of each command like you've already got with ANSI_WARNINGS, but all of that seems to me to be a bit of a hack. The fix I've included will work with any combination of triggers or print statements, so if another user wants to keep their triggers applying whatever business rules they've got setup, then the anonymization process will work for them in a predictable way.

jonstace-hsf commented:
Ok, so looking at the integration test failures (more fool me for not running them first), I can see why my change would break the restore db, dump db, etc stages. This change should only be applied to DML queries. I could just catch that exception, but that seems icky. What I'd like to do is change the __execute() and __db_execute() functions for MSSQL to be more like __execute_dml() and __execute_ddl(), and split the calls to the appropriate function depending upon what's being done, so that I can only do the Cursor.nextset() calls on DML queries.
Do that work? It does make it a larger change, but right now you can have really weird, unexpected behaviour on MSSQL where tables are silently only partially anonymized.

rwnx commented:
Thank you for your investigation on this! Yes, this seems really helpful. 
splitting to __execute_dml/ddl seems like a simple fix for now. 

Could you also add a changelog entry for this change? Happy to accept then!

jonstace-hsf commented:
I've now implemented this and run a full restore/anonymize/dump process in my test env, which happily runs. The integration tests are passing, which is great, and I've added an entry into CHANGELOG. 
Let me know if you need any further changes.

rwnx commented:
Fantastic work, Thank you so much!
